src/modules/mana/sagas.ts

```diff
@@ -484,6 +471,120 @@ function* handleConnectWalletSuccess(_action: ConnectWalletSuccessAction) {
   }
 }
 
+function formatImportWithdrawalError(msg: string) {
+  return `${IMPORT_WITHDRAWAL_REQUEST} - ${msg}`
+}
+
+export enum ImportWithdrawalErrors {
+  NOT_FOUND = 'notFound',
+  NOT_WITHDRAWAL = 'notWithdrawal',
+  NOT_OWN_TRANSACTION = 'notOwnTransaction',
+  ALREADY_PROCESSED = 'alreadyProcessed',
+}
+
+export const importWithdrawalErrors = {
+  notFound: formatImportWithdrawalError(ImportWithdrawalErrors.NOT_FOUND),
+  notWithdrawal: formatImportWithdrawalError(
+    ImportWithdrawalErrors.NOT_WITHDRAWAL
+  ),
+  notOwnTransaction: formatImportWithdrawalError(
+    ImportWithdrawalErrors.NOT_OWN_TRANSACTION
+  ),
+  alreadyProcessed: formatImportWithdrawalError(
+    ImportWithdrawalErrors.ALREADY_PROCESSED
+  ),
+  other: (msg: string) => formatImportWithdrawalError(msg),
+}
+
+function* handleImportWithdrawalRequest(action: ImportWithdrawalRequestAction) {
+  const {
+    payload: { txHash },
+  } = action
+
+  try {
+    const address: string | undefined = yield select(getAddress)
+    const networks: ReturnType<typeof getNetworks> = yield select(getNetworks)
+    const provider: Provider = yield call(() =>
+      getNetworkProvider(networks![Network.MATIC].chainId)
+    )
+
+    const transaction:
+      | { input: string; from: string }
+      | undefined = yield call(provider.send, 'eth_getTransactionByHash', [
+      txHash,
+    ])
+
+    if (!transaction) {
+      yield put(importWithdrawalFailure(importWithdrawalErrors.notFound))
+      return
+    }
+    const { input, from } = transaction
+
+    // hex for the "withdraw" method found in transaction.input
+    const method = '2e1a7d4d'
+    const methodIndex = input.indexOf(method)
+
+    if (methodIndex === -1) {
+      yield put(importWithdrawalFailure(importWithdrawalErrors.notWithdrawal))
+      return
+    }
+
+    if (from !== address && !input.includes(address!.slice(2))) {
+      yield put(importWithdrawalFailure(importWithdrawalErrors.notOwnTransaction))
+      return
+    }
+
+    const matic: MaticPOSClient = yield getMaticPOSClient()
+
+    let isProcessed: boolean
+
+    try {
+      isProcessed = yield call(() => matic.isERC20ExitProcessed(txHash))
+    } catch (e) {
+      isProcessed = false
+    }
+
+    if (isProcessed) {
+      yield put(
+        importWithdrawalFailure(importWithdrawalErrors.alreadyProcessed)
+      )
+      return
+    }
+
+    const methodEndIndex = methodIndex + method.length
+    const amountHex = input.slice(methodEndIndex, methodEndIndex + 64)
+    const amountDec = toBN(amountHex).div(toBN(1e18)).toNumber()
```

Comment:
Think it's necessary to use toBN here? Is this handled correctly?

Reply:
Just as an informational comment, the `.toNumber()` of `BN` can throw if the number is too big.. this is not a problem here since that number is bigger than the whole supply of MANA, and we are already within a try/catch, but keep this in mind if you use it within the `render` method of a component.. a few months ago someone made the marketplace inaccessible by putting a wearable on sale for a gazillion MANA, by interacting directly with the contract (it would be impossible to do from the UI) 😅